File: leetcode/dp/state-management/0740_delete_and_earn.py
# LC 740. Delete and Earn | Medium (State Management DP)
class Solution:
    def deleteAndEarn(self, nums):
        if not nums: return 0
        mx = max(nums); earn = [0]*(mx+1)
        for n in nums: earn[n] += n
        prev2, prev1 = 0, 0
        for i in range(mx+1): prev2, prev1 = prev1, max(prev1, prev2 + earn[i])
        return prev1

File: leetcode/dp/state-management/0740_delete_and_earn.cpp
// LC 740. Delete and Earn | Medium (State Management DP)
#include <vector>
#include <algorithm>
using namespace std;

class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int mx = *max_element(nums.begin(), nums.end());
        vector<int> earn(mx+1, 0);
        for (int n : nums) earn[n] += n;
        int p2 = 0, p1 = 0;
        for (int i = 0; i <= mx; i++) { int t = p1; p1 = max(p1, p2+earn[i]); p2 = t; }
        return p1;
    }
};
